File: golf2/apps/u2f_app/dcrypto.c
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <tock.h>
#include "include/dcrypto.h"


int last_error = 0;
int last_fault = 0;

const char* TOCK_DCRYPTO_ERRORS[] = {
                               "?",
                               "?",
                               "call stack overflow",
                               "loop stack underflow",
                               "loop stack overflow",
                               "data access",
                               "?",
                               "break",
                               "trap",
                               "?",
                               "fault",
                               "loop mod operand range",
                               "unknown"};

static const char* tock_dcrypto_fault_to_str(int fault) {
    if (fault <= TOCK_DCRYPTO_FAULT_UNKNOWN) {
      return TOCK_DCRYPTO_ERRORS[fault];
    } else {
      return "?";
    }
}

static void tock_dcrypto_run_done(int error,
                              int fault,
                              int unused __attribute__((unused)),
                              void *callback_args) {
  last_error = error;
  last_fault = fault;
  *(bool*)callback_args = true;
}



int tock_dcrypto_run(void* data, size_t datalen,
                     void* program, size_t programlen) {

  int ret = -1;
  bool run_done = false;

  ret = subscribe(HOTEL_DRIVER_DCRYPTO, TOCK_DCRYPTO_RUN_DONE,
                  tock_dcrypto_run_done, &run_done);
  if (ret < 0) {
    printf("Could not register dcrypto callback with kernel: %d\n", ret);
    return ret;
  }

  ret = allow(HOTEL_DRIVER_DCRYPTO, TOCK_DCRYPTO_ALLOW_DATA,
              data, datalen);
  if (ret < 0) {
    // This should only occur if application state is not available,
    // which means the driver is busy.
    printf("Could not give kernel access to dcrypto data: %d\n", ret);
    return TOCK_EBUSY;
  }

  ret = allow(HOTEL_DRIVER_DCRYPTO, TOCK_DCRYPTO_ALLOW_PROG,
              program, programlen);
  if (ret < 0) {
    // This should only occur if application state is not available,
    // which means the driver is busy.
    printf("Could not give kernel access to dcrypto program: %d\n", ret);
    return TOCK_EBUSY;
  }

  ret = command(HOTEL_DRIVER_DCRYPTO, TOCK_DCRYPTO_CMD_RUN, 0, 0);

  if (ret < 0) {
    printf("Could not invoke dcrypto program with command: %d\n", ret);
    return ret;
  }

  yield_for(&run_done);

  if (last_error != 0) {
    printf("\nDCRYPTO failed: %s (%i).\n", tock_dcrypto_fault_to_str(last_fault), last_fault);
    return last_fault;
  } else {
    return 0;
  }
}

File: h1_syscalls/src/dcrypto.rs
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use core::cell::Cell;
use h1::crypto::dcrypto::{Dcrypto, DcryptoClient, ProgramFault};
use kernel::{AppId, Callback, Driver, ReturnCode, Shared, AppSlice};
use kernel::common::cells::MapCell;

pub const DRIVER_NUM: usize = 0x40004;

pub struct App {
    program: Option<AppSlice<Shared, u8>>,
    data_buffer: Option<AppSlice<Shared, u8>>,
    callback: Option<Callback>,
}

impl Default for App {
    fn default() -> App {
        App {
            program: None,
            data_buffer: None,
            callback: None
        }
    }
}

pub struct DcryptoDriver<'a> {
    device: &'a dyn Dcrypto<'a>,
    app: MapCell<App>,
    busy: Cell<bool>,
}

impl<'a> DcryptoDriver<'a> {
    pub fn new(device: &'a mut dyn Dcrypto<'a>) -> DcryptoDriver<'a> {
        DcryptoDriver {
            device: device,
            app: MapCell::new(App::default()),
            busy: Cell::new(false),
       }
    }

    fn run_program(&self, app: &mut App, instruction: u32) -> ReturnCode {
        if app.data_buffer.is_none() || app.program.is_none() {
            return ReturnCode::ENOMEM;
        }

        let mut rval: ReturnCode;
        let data_slice = app.data_buffer.take().unwrap();
        let program_slice = app.program.take().unwrap();
        {
            // In user space, len is in bytes. For the device, however,
            // len is in terms of words, with partial words being truncated.
            // So divide by 4.
            let data = data_slice.as_ref();
            let data_len = data.len() / 4;
            let program = program_slice.as_ref();
            let program_len = program.len() / 4;

            rval = self.device.write_data(data, 0, data_len as u32);

            if rval == ReturnCode::SUCCESS {
                rval = self.device.write_instructions(program, 0, program_len as u32);
            }
        };
        app.data_buffer = Some(data_slice);
        app.program = Some(program_slice);

        if rval != ReturnCode::SUCCESS {
            return rval;
        }
        rval = self.device.call_imem(instruction);
        if rval != ReturnCode::SUCCESS {
            return rval;
        }
        ReturnCode::SUCCESS
    }
}

impl<'a> Driver for DcryptoDriver<'a> {
    fn subscribe(&self,
                 subscribe_num: usize,
                 callback: Option<Callback>,
                 _app_id: AppId,
    ) -> ReturnCode {
        match subscribe_num {
            0 => {
                self.app.map(|app| {
                    app.callback = callback;
                });
                ReturnCode::SUCCESS
            },
            _ => ReturnCode::ENOSUPPORT
        }
    }

    fn command(&self, command_num: usize, instruction: usize, _: usize, _: AppId) -> ReturnCode {
        match command_num {
            0 /* Check if present */ => ReturnCode::SUCCESS,
            1 /* run program */ => {
                if self.busy.get() {
                    ReturnCode::EBUSY
                } else {
                    self.app.map_or(ReturnCode::EBUSY, |app| {
                        self.busy.set(true);
                        self.run_program(app, instruction as u32)
                    })
                }
            }
            _ => ReturnCode::ENOSUPPORT,
        }
    }

    fn allow(&self, _: AppId,
             minor_num: usize,
             slice: Option<AppSlice<Shared, u8>>
    ) -> ReturnCode {
        match minor_num {
            0 => {
                // Data memory
                self.app
                    .map(|app_data| {
                        app_data.data_buffer = slice;
                        ReturnCode::SUCCESS
                    })
                    .unwrap_or(ReturnCode::FAIL)
            }
            1 => {
                // Input Buffer
                self.app
                    .map(|app_data| {
                        app_data.program = slice;
                        ReturnCode::SUCCESS
                    })
                    .unwrap_or(ReturnCode::FAIL)
            }
            _ => ReturnCode::ENOSUPPORT,
        }
    }
}

impl<'a> DcryptoClient<'a> for DcryptoDriver<'a> {
    fn execution_complete(&self, error: ReturnCode, fault: ProgramFault) {
        self.busy.set(false);
        self.app.map(move |app| {
            app.callback.map(|mut callback| {
                let mut data_slice = app.data_buffer.take().unwrap();
                {
                    let data = data_slice.as_mut();
                    // In user space, len is in bytes. For the device,
                    // however, len is in terms of words, with partial
                    // words being truncated.  So divide by 4.
                    let len = (data.len() / 4) as u32;
                    self.device.read_data(data, 0, len);
                    callback.schedule(usize::from(error), usize::from(fault), 0);
                }
                app.data_buffer = Some(data_slice);
            });
        });
    }

    fn reset_complete(&self, _error: ReturnCode) {
        panic!("ERROR: Dcrypto driver reset_complete invoked, but should never be called.");
    }

    fn secret_wipe_complete(&self, _error: ReturnCode) {
        panic!("ERROR: Dcrypto driver secret_wipe_complete invoked, but should never be called.");
    }


}
